File: Akshit_22b2531/week2/factory_machines.py
def solve():
    n, t = map(int, input().split())
    machines = list(map(int, input().split()))
    
    # Binary search for the minimum time
    left = 0
    right = 10**18  # Upper bound for the answer
    
    while left < right:
        mid = (left + right) // 2
        products = sum(mid // machine for machine in machines)
        
        if products >= t:
            right = mid
        else:
            left = mid + 1
    
    print(left)

if __name__ == "__main__":
    solve() 

File: week2/Factory Machines.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

bool canProduce(ll time, const vector<ll>& machines, ll t) {
    ll total = 0;
    for (ll k : machines) {
        total += time / k;
        if (total >= t) return true; 
    }
    return false;
}
int main() {
    cin.tie(nullptr);
    int n;
    ll t;
    cin >> n >> t;
    vector<ll> machines(n);
    for (int i = 0; i < n; ++i) {
        cin >> machines[i];
    }
    ll left = 1;
    ll right = 1e18;
    ll answer = -1;
    while (left <= right) {
        ll mid = left + (right - left) / 2;
        if (canProduce(mid, machines, t)) {
            answer = mid;
            right = mid - 1; 
        } else {
            left = mid + 1;
        }
    }
    cout << answer << "\n";
    return 0;
}
